lib/CL/devices/tbb/tbb.cc

```diff
@@ -0,0 +1,393 @@
+  volatile uint64_t total_cmd_exec_time;
+};
+
+void
+pocl_tbb_init_device_ops(struct pocl_device_ops *ops)
+{
+  pocl_basic_init_device_ops(ops);
+
+  ops->device_name = "tbb";
+
+  /* implementation that differs from basic */
+  ops->probe = pocl_tbb_probe;
+  ops->uninit = pocl_tbb_uninit;
+  ops->reinit = pocl_tbb_reinit;
+  ops->init = pocl_tbb_init;
+  ops->run = pocl_tbb_run;
+  ops->join = pocl_tbb_join;
+  ops->submit = pocl_tbb_submit;
+  ops->notify = pocl_tbb_notify;
+  ops->broadcast = pocl_broadcast;
+  ops->flush = pocl_tbb_flush;
+  ops->wait_event = pocl_tbb_wait_event;
+  ops->notify_event_finished = pocl_tbb_notify_event_finished;
+  ops->notify_cmdq_finished = pocl_tbb_notify_cmdq_finished;
+  ops->update_event = pocl_tbb_update_event;
+  ops->free_event_data = pocl_tbb_free_event_data;
+  ops->build_hash = pocl_tbb_build_hash;
+
+  ops->init_queue = pocl_tbb_init_queue;
+  ops->free_queue = pocl_tbb_free_queue;
+}
+
+char*
+pocl_tbb_build_hash (cl_device_id device)
+{
+  char* res = reinterpret_cast<char*>(calloc(1000, sizeof(char)));
+#ifdef KERNELLIB_HOST_DISTRO_VARIANTS
+  char *name = get_llvm_cpu_name ();
+  snprintf (res, 1000, "tbb-%s-%s", HOST_DEVICE_BUILD_HASH, name);
+  POCL_MEM_FREE (name);
+#else
+  snprintf (res, 1000, "tbb-%s", HOST_DEVICE_BUILD_HASH);
+#endif
+  return res;
+}
+
+unsigned int
+pocl_tbb_probe (struct pocl_device_ops *ops)
+{
+  int env_count = pocl_device_get_env_count(ops->device_name);
+  /* Env was not specified, default behavior was to use 1 tbb device */
+  if (env_count < 0)
+    return 1;
+
+  return env_count;
+}
+
+static cl_device_partition_property tbb_partition_properties[2]
+    = { CL_DEVICE_PARTITION_EQUALLY, CL_DEVICE_PARTITION_BY_COUNTS };
+
+char scheduler_initialised = 0;
+
+cl_int
+pocl_tbb_init (unsigned j, cl_device_id device, const char* parameters)
+{
+  struct data *d;
+  cl_int ret = CL_SUCCESS;
+  int err;
+
+  d = (struct data *) calloc (1, sizeof (struct data));
+  if (d == NULL)
+    return CL_OUT_OF_HOST_MEMORY;
+
+  d->current_kernel = NULL;
+  device->data = d;
+
+  pocl_init_default_device_infos (device);
+  device->extensions = HOST_DEVICE_EXTENSIONS;
+
+  device->on_host_queue_props
+      = CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE | CL_QUEUE_PROFILING_ENABLE;
+
+  /* hwloc probes OpenCL device info at its initialization in case
+     the OpenCL extension is enabled. This causes to printout 
+     an unimplemented property error because hwloc is used to
+     initialize global_mem_size which it is not yet. Just put 
+     a nonzero there for now. */
+  device->global_mem_size = 1;
+  err = pocl_topology_detect_device_info (device);
+  if (err)
+    ret = CL_INVALID_DEVICE;
+
+  /* device->max_compute_units was set up by topology_detect,
+     but we use the TBB library (result should be the same).
+     task_area initialization is optional and max_concurrency
+     can be retrieved without prior initialization. */
+  tbb::task_arena ta;
+  //ta.initialize();
```

Comment:
artefact?

Reply:
I left it there intentionally. Initialisation is optional as stated in the comment above.